contracts/src/token/erc20/extensions/wrapper.rs

```diff
@@ -0,0 +1,279 @@
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc20Wrapper {}
+
+impl IERC20Wrapper for Erc20Wrapper {
+    type Error = Error;
+
+    fn underlying(&self) -> Address {
+        self.underlying_address.get()
+    }
+
+    fn deposit_for(
+        &mut self,
+        account: Address,
+        value: U256,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Error> {
+        let underlined_token = self.underlying_address.get();
+        let sender = msg::sender();
+
+        if sender == contract::address() {
+            return Err(Error::InvalidReceiver(ERC20InvalidReceiver {
+                receiver: account,
+            }));
+        }
+
+        if account == contract::address() {
+            return Err(Error::InvalidSender(ERC20InvalidSender {
+                sender: contract::address(),
+            }));
+        }
+
+        self.safe_erc20.safe_transfer_from(
+            underlined_token,
+            sender,
+            contract::address(),
+            value,
+        )?;
+        erc20._mint(account, value)?;
+        Ok(true)
+    }
+
+    fn withdraw_to(
+        &mut self,
+        account: Address,
+        value: U256,
+        erc20: &mut Erc20,
+    ) -> Result<bool, Error> {
+        let underlined_token = self.underlying_address.get();
+        if account == contract::address() {
+            return Err(Error::InvalidReceiver(ERC20InvalidReceiver {
+                receiver: account,
+            }));
+        }
+        erc20._burn(account, value)?;
+        self.safe_erc20.safe_transfer(underlined_token, account, value)?;
+        Ok(true)
+    }
+}
+
+impl Erc20Wrapper {
+    /// Mints wrapped tokens to cover any underlying tokens that might have been
+    /// mistakenly transferred or acquired through rebasing mechanisms.
+    ///
+    /// This is an internal function that can be exposed with access control if
+    /// required.
+    ///
+    /// Arguments:
+    ///
+    /// * `&mut self` - Write access to the contract's state.
+    /// * `account` - The account to mint tokens to.
+    /// * `erc20` - A mutable reference to the Erc20 contract.
+    ///
+    /// # Errors
+    ///
+    /// If the external call for balance of fails , then the error
+    /// [`Error::InvalidAsset`] is returned.
+    pub fn _recover(
+        &mut self,
+        account: Address,
+        erc20: &mut Erc20,
+    ) -> Result<U256, Error> {
+        let token = IErc20Solidity::new(self.underlying());
+        let value = token
+            .balance_of(Call::new_in(self), contract::address())
+            .map_err(|_| InvalidAsset { asset: contract::address() })?;
+        erc20._mint(account, value)?;
+        Ok(U256::from(value))
+    }
+}
+
+// TODO: Add missing tests once `motsu` supports calling external contracts.
```

Comment:
We should have motsu version that support calling external contracts tomorrow :)
It is already merged into main branch of mostu. You can implement these test cases.